### PR TITLE
[AIRFLOW-3411]  create openfaas hook

### PR DESCRIPTION
**Pull request by tal181**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following https://issues.apache.org/jira/browse/AIRFLOW-3411  issues and references them in the PR title. 
AIRFLOW-3411 : create openfaas hook
### Description
Adding a hook for openfaas

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
tests/contrib/hooks/test_openfaas_hook.py
### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

**Comment by tal181** [comment]
ping @kaxil

File: airflow/contrib/hooks/openfaas_hook.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    def get_conn(self):
+        conn = self.get_connection(self.conn_id)
+        return conn
+
+    def deploy_function(self, overwrite_function_if_exist, body):
+        if overwrite_function_if_exist:
+            self.log.info("function already exist " + self.function_name + " going to update")
+            self.update_function(body)
+        else:
+            url = self.get_conn().host + self.DEPLOY_FUNCTION
```
Comment (kaxil) [inline airflow/contrib/hooks/openfaas_hook.py:60]: Need to add better handling, what are the required fields in the connection. We don't have any connection for `open_faas_default`.

File: airflow/contrib/hooks/openfaas_hook.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            self.log.info("deploying function " + url)
```
Comment (kaxil) [inline airflow/contrib/hooks/openfaas_hook.py]: First alphabet of anything in Logging should be capital as a convention

File: airflow/contrib/hooks/openfaas_hook.py [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+class OpenFaasHook(BaseHook):
+    """
+    Interact with Openfaas to query, deploy, invoke and update function
+
+    :param function_name: Name of the function, Defaults to None
+    :type query: str
+    :param conn_id: openfass connection to use, Defaults to open_faas_default
+    :type conn_id: str
+    """
```
Comment (kaxil) [inline airflow/contrib/hooks/openfaas_hook.py:36]: Add an example of what needs to be in open-faas connection

**Review by kaxil — CHANGES_REQUESTED** [review]

**Comment by tal181** [comment]
ping @kaxil

File: airflow/contrib/hooks/openfaas_hook.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+
+class OpenFaasHook(BaseHook):
+    """
+    Interact with Openfaas to query, deploy, invoke and update function
+
+    :param function_name: Name of the function, Defaults to None
+    :type query: str
+    :param conn_id: openfass connection to use, Defaults to open_faas_default
+    for example host : http://openfaas.faas.com, Conn Type : Http
```
Comment (kaxil) [inline airflow/contrib/hooks/openfaas_hook.py]: Need 1 indent over here

**Review by kaxil — CHANGES_REQUESTED** [review]

File: airflow/contrib/hooks/openfaas_hook.py (hunk 4 above)
Comment (tal181) [reply airflow/contrib/hooks/openfaas_hook.py]: done

**Comment by tal181** [comment]
ping @kaxil

**Review by kaxil — APPROVED** [review]

File: airflow/contrib/hooks/openfaas_hook.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+from airflow.hooks.base_hook import BaseHook
+import requests
+from airflow import AirflowException
+
+OK_STATUS_CODE = 202
+
+
+class OpenFaasHook(BaseHook):
+    """
+    Interact with Openfaas to query, deploy, invoke and update function
```
Comment (alexellis) [inline airflow/contrib/hooks/openfaas_hook.py:29]: Please could you use the correct trademark of `OpenFaaS`?

File: airflow/contrib/hooks/openfaas_hook.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+    :param function_name: Name of the function, Defaults to None
+    :type query: str
+    :param conn_id: openfass connection to use, Defaults to open_faas_default
```
Comment (alexellis) [inline airflow/contrib/hooks/openfaas_hook.py:33]: This is a typo, it's `openfaas` not `openfass`

Comment (alexellis) [inline airflow/contrib/hooks/openfaas_hook.py:33]: Please do not split the trademark of `openfaas`, this should be `openfaas_default`

File: docs/code.rst [hunk 7]
```diff
@@ -442,6 +442,7 @@ Community contributed hooks
 .. autoclass:: airflow.contrib.hooks.jenkins_hook.JenkinsHook
 .. autoclass:: airflow.contrib.hooks.jira_hook.JiraHook
 .. autoclass:: airflow.contrib.hooks.mongo_hook.MongoHook
+.. autoclass:: airflow.contrib.hooks.openfaas_hook.OpenFaasHook
```
Comment (alexellis) [inline docs/code.rst:445]: Ideally write `OpenFaaS`

File: tests/contrib/hooks/test_openfaas_hook.py [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+        with self.assertRaises(AirflowException) as context:
+            self.hook.invoke_async_function({})
+        self.assertIn('failed to invoke function', str(context.exception))
+
+    @mock.patch.object(BaseHook, 'get_connection')
+    @requests_mock.mock()
+    def test_invoke_async_function_true(self, mock_get_connection, m):
+        m.post("http://open-faas.io" + self.INVOKE_ASYNC_FUNCTION + FUNCTION_NAME, json=self.mock_response,
+               status_code=202)
+        mock_connection = Connection(host="http://open-faas.io")
```
Comment (alexellis) [inline tests/contrib/hooks/test_openfaas_hook.py:110]: Might be better to use a fictitious URL rather than one that looks similar to the official website?

**Pull request merged**